### PR TITLE
Introduce the 15-wires PLONK code

### PR DESCRIPTION
**Pull request by mimoo**
This PR builds on https://github.com/o1-labs/proof-systems/pull/126 to introduce the 15-wires PLONK construction to the codebase.

In order of importance it:

1. [introduces a protocol change to always leave room for the zero knowledge polynomial](https://github.com/o1-labs/proof-systems/pull/137/commits/410174ee1f0a857ffc32a252577d45165b3e64e0)
1. fixes the generic gate
1. fixes the poseidon gate (closes https://github.com/o1-labs/proof-systems/issues/136)
1. adds a few of assertions
1. adds a number of tests
1. adds a number of (rustdoc) comments
1. refactors some places to make the code cleaner
2. merged master into this branch AND updated the code to use most recent arkworks
1. ran `cargo fmt`

What's the minimum that needs to be done to merge this:

- [ ] someone makes sure that the [zero polynomial change](https://github.com/o1-labs/proof-systems/pull/137/commits/410174ee1f0a857ffc32a252577d45165b3e64e0) is not a stupid change
- [x] code needs to be updated to match the [arkworks branch](https://github.com/o1-labs/proof-systems/pull/124)

I propose that once this is cleaner/reviewed we merge it, and then work on a) more tests to make sure all gates work, etc. and b) OCaml bindings.

### CODE REVIEW

**Comment by mimoo** [comment]
merged master into this branch, and updated code to arkworks.

**Comment by mimoo** [comment]
for some reason, after the update to arkworks, the poseidon test doesn't pass anymore. Investigating.

**Comment by mimoo** [comment]
fixed

File: circuits/plonk-15-wires/src/domains.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    pub fn create(n: usize) -> Option<Self> {
+        let n = Domain::<F>::compute_size_of_domain(n)?;
+
+        let d1 = Domain::<F>::new(n)?;
+        let d4 = Domain::<F>::new(4 * n)?;
+        let d8 = Domain::<F>::new(8 * n)?;
+
+        // ensure the relationship between the three domains in case the library's behavior changes
+        assert!(d4.group_gen.pow(&[4]) == d1.group_gen);
+        assert!(d8.group_gen.pow(&[2]) == d4.group_gen);
```
Comment (mimoo) [inline circuits/plonk-15-wires/src/domains.rs:25]: added these asserts to ensure that there's a relationship between our different domains.

File: circuits/plonk-15-wires/src/gates/poseidon.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+/// returns the columns (as a range) that are used in this round.
+pub const fn round_to_cols(i: usize) -> Range<usize> {
+    let slot = STATE_ORDER[i];
+    let start = slot * SPONGE_WIDTH;
+    start..(start + SPONGE_WIDTH)
+}
+
+/// Apply an MDS matrix to a state
+fn apply_mds<F: FftField>(mds: &[Vec<F>], state: &[F]) -> Vec<F> {
+    assert!(mds.len() == state.len());
```
Comment (mimoo) [inline circuits/plonk-15-wires/src/gates/poseidon.rs:55]: added that, as previously it was possible to set a state larger than the width

File: circuits/plonk-15-wires/src/gates/poseidon.rs [hunk 3]
```diff
@@ -43,6 +43,51 @@ impl<F: FftField> CircuitGate<F> {
...
+    /// For that, you need to pass:
+    /// - the index of the first `row`
+    /// - the first and last rows' wires (because they are used in the permutation)
+    /// - the round constants
+    /// The function returns a set of gates, as well as the next pointer to the circuit (next empty absolute row)
+    pub fn create_poseidon_gadget(
+        // the absolute row in the circuit
+        row: usize,
+        // first and last row of the poseidon circuit (because they are used in the permutation)
+        first_and_last_row: [GateWires; 2],
```
[the remark this answers is not in the document]
Comment (mimoo) [reply circuits/plonk-15-wires/src/gates/poseidon.rs:78]: yep that could be better.

**Comment by mimoo** [comment]
forced push to fix the "remove me" commit. Will rebase next.

**Comment by mimoo** [comment]
merged master, and added small fix to CI

**Comment by spartucus** [comment]
Hi @mimoo, are there any papers or docs about kimchi?

**Comment by mimoo** [comment]
there's https://o1-labs.github.io/mina-book/ but it's heavy work in progress

**Comment by spartucus** [comment]
Thanks :)

**Pull request merged**